docker/scripts/setup_env

```diff
@@ -5,31 +5,68 @@ export LC_ALL=en_US.UTF-8
-  DATABASE_ADAPTER=${DATABASE_ADAPTER:-postgresql}
-  DATABASE_HOST=${DATABASE_HOST:-${POSTGRES_PORT_5432_TCP_ADDR}}
-  DATABASE_PORT=${DATABASE_PORT:-${POSTGRES_PORT_5432_TCP_PORT}}
-  DATABASE_ENCODING=utf8
+if ! whoami &> /dev/null; then
+    export HOME=/app
+    if [ -w /etc/passwd ]; then
+        echo "${USER_NAME:-default}:x:$(id -u):0:${USER_NAME:-default} user:${HOME}:/sbin/nologin" >> /etc/passwd
+    fi
 fi
 
-USE_GRAPHVIZ_DOT=${USE_GRAPHVIZ_DOT:-${USE_GRAPHVIZ_DOT:-dot}}
+# Do we have a Rails secret yet?
+# If not, generate one, but persist it for the container's lifetime.
+if [ -z "$APP_SECRET_TOKEN" ] && grep -q ^APP_SECRET_TOKEN=REPLACE_ME_NOW .env.example; then
+    echo 'Generating random APP_SECRET_TOKEN.'
+    secret=$(dd if=/dev/urandom bs=36 count=1 | openssl base64)
+    sed -i "s:REPLACE_ME_NOW!:$secret:" .env.example
+fi
+
+# is a mysql or postgresql database linked?
+# requires that the mysql or postgresql containers have exposed
+# port 3306 and 5432 respectively.
+if [[ -n "${MYSQL_PORT_3306_TCP_ADDR}" || ("${DATABASE_ADAPTER}" == "mysql2" && -n "${DATABASE_HOST}") ]]; then
+    DATABASE_ADAPTER=${DATABASE_ADAPTER:-mysql2}
+    DATABASE_HOST=${DATABASE_HOST:-${MYSQL_PORT_3306_TCP_ADDR}}
+    DATABASE_PORT=${DATABASE_PORT:-${MYSQL_PORT_3306_TCP_PORT:-3306}}
+    DATABASE_ENCODING=${DATABASE_ENCODING:-utf8mb4}
+elif [[ -n "${POSTGRES_PORT_5432_TCP_ADDR}" || ("${DATABASE_ADAPTER}" == "postgresql" && -n "${DATABASE_HOST}") ]]; then
+    DATABASE_ADAPTER=${DATABASE_ADAPTER:-postgresql}
+    DATABASE_HOST=${DATABASE_HOST:-${POSTGRES_PORT_5432_TCP_ADDR}}
+    DATABASE_PORT=${DATABASE_PORT:-${POSTGRES_PORT_5432_TCP_PORT:-5432}}
+    DATABASE_ENCODING=utf8
+else
+    START_MYSQL=${START_MYSQL:-true}
+fi
+
+USE_GRAPHVIZ_DOT=${HUGINN_USE_GRAPHVIZ_DOT:-${USE_GRAPHVIZ_DOT:-dot}}
+DATABASE_HOST=${HUGINN_DATABASE_HOST:-${DATABASE_HOST:-localhost}}
+DATABASE_PORT=${HUGINN_DATABASE_PORT:-${DATABASE_PORT}}
+DATABASE_ENCODING=${HUGINN_DATABASE_ENCODING:-${DATABASE_ENCODING}}
+DATABASE_PASSWORD=${HUGINN_DATABASE_PASSWORD:-${DATABASE_PASSWORD:-password}}
+DATABASE_NAME=${HUGINN_DATABASE_NAME:-${DATABASE_NAME:-huginn_production}}
+RAILS_ENV=${HUGINN_RAILS_ENV:-${RAILS_ENV:-production}}
 
 # Default to the environment variable values set in .env.example
 IFS="="
-grep = /app/.env.example | sed -e 's/^#\([^ ]\)/\1/' | grep -v -e '^#' | \
+sed -n -r -e 's/^#?([A-Za-z0-9_]+ *=.*)/\1/p' /app/.env.example | \
   while read var value ; do
-    eval "echo \"$var=\${$var:-\${HUGINN_$var-\$value}}\""
-  done | grep -v -e ^= > /app/.env
+    eval "echo \"$var=\${HUGINN_$var:-\${$var:-\$value}}\""
+  done > /app/.env
 
-eval "echo PORT=${PORT:-${PORT:-3000}}" >> .env
-eval "echo RAILS_ENV=${RAILS_ENV:-${RAILS_ENV:-production}}" >> .env
-eval "echo RAILS_LOG_TO_STDOUT=true" >> .env
-eval "echo RAILS_SERVE_STATIC_FILES=true" >> .env
+PORT=${HUGINN_PORT:-${PORT:-3000}}
+echo "PORT=$(echo $PORT|awk -F ':' '{print $NF}')" >> .env
```

Comment:
> It was actually trying to handle an unintended case. Everything defaults to use port 3000, so even if PORT or HUGINN_PORT is not specified, things will work fine (provided a mechanism such as -p 3000:3000 is set up to forward it). It was natural to call the service "huginn", but that environment variable messes the setup.

`HUGINN_PORT` gets used as `PORT` due to the `HUGINN_` variable prefix we are using in the past? I think `HUGINN_PORT` was never documented so we could exclude it from that logic to get rid of this line.

Reply:
A potential complication is that `PORT` is also in `.env.example`, and there is logic to check every variable in that file: `eval "echo \"$var=\${HUGINN_$var:-\${$var:-\$value}}\""`, so will have to either make a special case there or delete `PORT` from `.env.example`. None of them seems ideal either..

Reply:
That is true, I don't particularly like having provider specific code in there. Did you find documentation why kubernetes is setting `<SERVICE>_PORT` to a tcp connection string instead of an actual port? [This](https://kubernetes.io/docs/concepts/containers/container-environment-variables/#cluster-information) reads like it should only pass a port .

Reply:
[It](https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables) looks like that Kubernetes actually inherited this "feature" from [Docker](https://docs.docker.com/engine/userguide/networking/default_network/dockerlinks/) itself.

Reply:
Ah that makes sense, I never noticed the `*_PORT` variable. I guess it makes sense to handle that case then.